themes/default/content/blog/troubleshooting-builds-with-diagnostics/index.md

```diff
@@ -0,0 +1,154 @@
+
+```go
+export function leakedPromises(): [Set<Promise<any>>, string] {
+    const leaked = leakCandidates;
+    const promisePlural = leaked.size === 0 ? "promise was" : "promises were";
+    const message = leaked.size === 0 ? "" :
+        `The Pulumi runtime detected that ${leaked.size} ${promisePlural} still active\n` +
+        "at the time that the process exited. There are a few ways that this can occur:\n" +
+        "  * Not using `await` or `.then` on a Promise returned from a Pulumi API\n" +
+        "  * Introducing a cyclic dependency between two Pulumi Resources\n" +
+        "  * A bug in the Pulumi Runtime\n" +
+        "\n" +
+        "Leaving promises active is probably not what you want. If you are unsure about\n" +
+        "why you are seeing this message, re-run your program "
+            + "with the `PULUMI_DEBUG_PROMISE_LEAKS`\n" +
+        "environment variable. The Pulumi runtime will then print out additional\n" +
+        "debug information about the leaked promises.";
+
+    if (debugPromiseLeaks) {
+        for (const leak of leaked) {
+            console.error("Promise leak detected:");
+            console.error(promiseDebugString(leak));
+        }
+    }
+
+    leakCandidates = new Set<Promise<any>>();
+    return [leaked, message];
+}
+```
+
+The variable causes Pulumi to dump the context, stack trace, and error for the leak. It's very rare you'll need to do this kind of debugging, but in case you do:
+
+```bash
+PULUMI_DEBUG_PROMISE_LEAKS=true pulumi up -v=11
+```
+
+### Data return
+
+Those flags, though, don't tell the CLI where to send the data for you to examine. You might be working on a system that doesn't allow you to write logs anywhere, you might be streaming that data to another platform, or you might be able to write to a file to examine or parse later. If you want to write the diagnostic data to stdout/stderr, you need to add the `--logtostderr` flag:
+
+```bash
+TF_LOG=DEBUG pulumi up -v=9 --logflow --logtostderr
+```
+
+If you want to write that diagnostic data to a file and not to stdout/stderr, you can use your standard output pipe to point the data to a file (in this example, we're appending to any existing logfile):
+
+```bash
+TF_LOG=TRACE pulumi up -v=11 --logflow --logtostderr &>> pulumi_log.txt
+```
+
+Alternatively, you could stream that diagnostic data to stdout/stderr *and* write it to a file with a slight change to that line:
+
+```bash
+TF_LOG=TRACE pulumi up -v=11 --logflow --logtostderr 2>&1 | tee -a pulumi_log.txt
+```
+
+## Understanding data
+
+While it's great to get the data, it's not useful unless we can understand what the troubleshooting data actually means. If you were to run these commands, you would get what might appear to be a bunch of unhelpful data:
```

Comment:
I feel like here - and in a number of other places - we could rephrase to be more positive - eg. 

```suggestion
While it's great to get the data, it's not useful unless we can understand what the troubleshooting data actually means. If you were to run these commands, you would get a ton of valuable information about what happened during the deployment, but how do we identify the pieces of information we need to troubleshoot our specific problem?:
```